### PR TITLE
Implemented Backup by a given distance 

### PR DESCRIPTION
**Pull request by mhpanah**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #380 |
| Primary OS tested on | Ubuntu 18.04|
| Robotic platform tested on |gazebo simulation of  Turtlebot |

---

## Description of contribution in a few bullet points

- This method is not based on any control theory and overshoot is expected to happen

- If the given distance is positive, robot moves in +x direction. If the given distance is negative, robot moves in -x direction

- Deleted time based method, since its no longer required

- Added loop rate to motion_primitives

---

## Future work that may be required in bullet points

---

I'd like to request maintainer: <blank> to review this PR.

### CODE REVIEW

File: nav2_motion_primitives/src/back_up.cpp [hunk 1]
```diff
@@ -66,46 +65,34 @@ nav2_tasks::TaskStatus BackUp::onCycleUpdate(nav2_tasks::BackUpResult & result)
...
-  auto current_time = std::chrono::system_clock::now();
-  if (current_time - start_time_ >= 3s) {
-    // Stop the robot
-    cmd_vel.linear.x = 0.0;
-    robot_->sendVelocity(cmd_vel);
+  double diff_x = initial_pose_->pose.pose.position.x - current_odom_pose->pose.pose.position.x;
+  double diff_y = initial_pose_->pose.pose.position.y - current_odom_pose->pose.pose.position.y;
+  double distance = sqrt(diff_x * diff_x + diff_y * diff_y);
 
+  if (distance > abs(command_x_)) {
```
Comment (crdelsey) [inline nav2_motion_primitives/src/back_up.cpp]: Can we change this to `>=`? That way if someone sends a command distance of 0, we'll succeed immediately.

File: nav2_motion_primitives/src/back_up.cpp [hunk 2]
```diff
@@ -66,46 +65,34 @@ nav2_tasks::TaskStatus BackUp::onCycleUpdate(nav2_tasks::BackUpResult & result)
...
+nav2_tasks::TaskStatus BackUp::controlledBackup()
 {
-  // Output control command
-  geometry_msgs::msg::Twist cmd_vel;
+  auto current_odom_pose = std::make_shared<nav_msgs::msg::Odometry>();
 
-  // TODO(orduno): #423 assuming robot was moving fwd when it got stuck
-  //               fixed speed
-  cmd_vel.linear.x = -0.025;
+  if (!robot_->getOdometry(current_odom_pose)) {
```
Comment (crdelsey) [inline nav2_motion_primitives/src/back_up.cpp]: It seems unnecessary for getOdometry to take a SharedPtr. It should just take a simple pointer. That way the client code could use a simple local variable instead. That is, `current_odom_pose` could be a local variable, which seems the most desirable.

File: nav2_motion_primitives/src/back_up.cpp [hunk 3]
```diff
@@ -45,19 +43,20 @@ nav2_tasks::TaskStatus BackUp::onRun(const nav2_tasks::BackUpCommand::SharedPtr
     RCLCPP_INFO(node_->get_logger(), "Backing up in Y and Z not supported, "
       "will only move in X.");
   }
+  command_x_ = command->x;
+  if (!robot_->getOdometry(initial_pose_)) {
+    RCLCPP_ERROR(node_->get_logger(), "initial robot odom pose is not available.");
+    return nav2_tasks::TaskStatus::FAILED;
+  }
 
   RCLCPP_INFO(node_->get_logger(), "Currently only supported backing up by a fixed distance");
```
Comment (orduno) [inline nav2_motion_primitives/src/back_up.cpp]: Well, it is now supported :) we can remove this line.

File: nav2_motion_primitives/src/back_up.cpp (hunk 2 above)
Comment (mhpanah) [reply nav2_motion_primitives/src/back_up.cpp]: Will address this on a different PR to also change getCurrentPose and getGlobalLocalizationPose.

File: nav2_motion_primitives/src/back_up.cpp [hunk 4]
```diff
@@ -45,19 +43,20 @@ nav2_tasks::TaskStatus BackUp::onRun(const nav2_tasks::BackUpCommand::SharedPtr
     RCLCPP_INFO(node_->get_logger(), "Backing up in Y and Z not supported, "
       "will only move in X.");
   }
+  command_x_ = command->x;
+  if (!robot_->getOdometry(initial_pose_)) {
```
Comment (orduno) [inline nav2_motion_primitives/src/back_up.cpp]: A suggestion for the `robot` class, maybe we should have an `isOdometryAvailable` method, here we could use it like this:

```cpp
if (!robot_->isOdometryAvailable())
```

Comment (orduno) [reply nav2_motion_primitives/src/back_up.cpp]: Oh, but I also see you actually use the `initial_pose_` never mind.

**Pull request merged**